File: scripts/aggregate.py
#!/usr/bin/env python3
"""Aggregate mined Jira JSONL into app/data/edges.json and app/data/pod_stats.json."""
import json
import math
from collections import Counter, defaultdict
from datetime import datetime
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
DATA = ROOT / "data"
OUT = ROOT / "app" / "data"


def load_jsonl(name):
    path = DATA / name
    if not path.exists():
        return []
    return [json.loads(l) for l in path.read_text().splitlines() if l.strip()]


# whitespace variants seen in Jira vs the org directory
POD_ALIASES = {"Moose Factory": "MooseFactory"}


def pod_of(rec):
    p = (rec.get("pod") or "").strip()
    return POD_ALIASES.get(p, p) or None


def main():
    links = load_jsonl("issues_links.jsonl")
    cycles = load_jsonl("cycle_times.jsonl")
    wip = load_jsonl("wip.jsonl")
    key_pods = load_jsonl("key_pods.jsonl") + load_jsonl("key_pods_extra.jsonl")

    pod_by_key = {}
    for rec in cycles + links + key_pods:
        p = pod_of(rec)
        if p:
            pod_by_key[rec["key"]] = p

    # deduplicated (blocker, blocked) pairs -> pod edges
    pairs = set()
    for rec in links:
        for k in rec.get("blocks", []):
            pairs.add((rec["key"], k))
        for k in rec.get("blockedBy", []):
            pairs.add((k, rec["key"]))
    edge_counts = Counter()
    unknown = 0
    for blocker, blocked in pairs:
        pa, pb = pod_by_key.get(blocker), pod_by_key.get(blocked)
        if not pa or not pb:
            unknown += 1
            continue
        if pa != pb:
            edge_counts[(pa, pb)] += 1
    # deterministic order (set iteration isn't stable across runs): count desc,
    # then names — so `make data` reproduces byte-identical output
    edges = [{"from": a, "to": b, "count": c}
             for (a, b), c in sorted(edge_counts.items(), key=lambda kv: (-kv[1], kv[0][0], kv[0][1]))]
    OUT.joinpath("edges.json").write_text(json.dumps(edges, indent=1))

    # cycle times per pod. created->resolved overstates flow time for tickets
    # that aged in the backlog, so: drop epics, drop >180d records (bulk
    # backlog-cleanup artifacts), winsorize the rest at the pod's p95.
    EXCLUDE_TYPES = {"Epic", "Parent Epic"}
    MAX_DAYS = 180
    by_pod = defaultdict(list)
    excluded = 0
    for rec in cycles:
        p = pod_of(rec)
        if not p or not rec.get("resolved") or not rec.get("created"):
            continue
        if rec.get("issuetype") in EXCLUDE_TYPES:
            excluded += 1
            continue
        try:
            c = datetime.fromisoformat(rec["created"])
            r = datetime.fromisoformat(rec["resolved"])
        except ValueError:
            continue
        days = max((r - c).total_seconds() / 86400, 0.25)
        if days > MAX_DAYS:
            excluded += 1
            continue
        by_pod[p].append(days)
    for p, xs in by_pod.items():
        xs.sort()
        cap = xs[int(0.95 * (len(xs) - 1))]
        by_pod[p] = [min(x, cap) for x in xs]
    print(f"excluded {excluded} records (epics or >{MAX_DAYS}d backlog artifacts)")

    wip_counts = Counter(pod_of(r) for r in wip if pod_of(r))

    def pct(xs, q):
        xs = sorted(xs)
        idx = (q / 100) * (len(xs) - 1)
        lo, hi = math.floor(idx), math.ceil(idx)
        return xs[lo] + (xs[hi] - xs[lo]) * (idx - lo)

    stats = {}
    for p, xs in by_pod.items():
        logs = [math.log(x) for x in xs]
        mu = sum(logs) / len(logs)
        sigma = math.sqrt(sum((l - mu) ** 2 for l in logs) / len(logs)) or 0.1
        stats[p] = {
            "resolved_count_180d": len(xs),
            "cycle_time_days": {
                "p50": round(pct(xs, 50), 2),
                "p85": round(pct(xs, 85), 2),
                "mean": round(sum(xs) / len(xs), 2),
            },
            "lognormal": {"mu": round(mu, 4), "sigma": round(sigma, 4)},
            "wip_count": wip_counts.get(p, 0),
            "throughput_per_week": round(len(xs) / 26, 2),
        }
    # pods with WIP but nothing resolved still deserve an entry
    for p, w in wip_counts.items():
        stats.setdefault(p, {
            "resolved_count_180d": 0,
            "cycle_time_days": {"p50": 7.0, "p85": 17.0, "mean": 10.0},
            "lognormal": {"mu": math.log(7), "sigma": 0.9},
            "wip_count": w,
            "throughput_per_week": 0.0,
        })
    OUT.joinpath("pod_stats.json").write_text(json.dumps(stats, indent=1))

    print(f"issues with links: {len(links)}, cycle records: {len(cycles)}, wip records: {len(wip)}")
    print(f"dedup link pairs: {len(pairs)}, dropped (unknown pod): {unknown}")
    print(f"cross-pod edges: {len(edges)}, pods with stats: {len(stats)}")
    print("top 10 edges:")
    for e in edges[:10]:
        print(f"  {e['from']} -> {e['to']}: {e['count']}")
    print("top 5 WIP:", wip_counts.most_common(5))


if __name__ == "__main__":
    main()

File: server/jira/aggregate.go
// Package jira fetches issues from Jira Cloud and aggregates them into the org
// snapshot documents the engine and Observe consume. The aggregation mirrors
// the offline scripts/aggregate.py so a live import matches the mined baseline.
package jira

import (
	"math"
	"sort"
	"time"
)

// Issue is the minimal shape the aggregation needs from a Jira issue.
type Issue struct {
	Key       string
	Pod       string     // customfield_10026 value (the pod/team)
	IssueType string     // "Epic", "Story", …
	StatusCat string     // statusCategory key: new | indeterminate | done
	ParentKey string     // parent epic's key, "" if none
	Created   time.Time  // zero if unknown
	Resolved  *time.Time // nil if still open
	Blocks    []string   // keys this issue blocks
	BlockedBy []string   // keys that block this issue
}

// WIP counting modes: an in-progress epic and its in-progress children both
// represent the "same" work if counted naively, so a snapshot picks one rule
// at import time.
const (
	// WipModeLeaf counts every non-epic in-progress issue, regardless of
	// whether it has a parent epic. Epics themselves (containers, not units of
	// flow) are never counted. This is the accurate view of concurrent work:
	// an epic with 5 active children is wip=5, not wip=1.
	WipModeLeaf = "leaf"
	// WipModeEpicOrParentless counts an in-progress epic as one unit and does
	// not separately count its children; parentless stories/tasks still count
	// individually.
	WipModeEpicOrParentless = "epic_or_parentless"
)

// PodStat mirrors one entry of pod_stats.json.
type PodStat struct {
	ResolvedCount180d int        `json:"resolved_count_180d"`
	CycleTimeDays     CycleTimes `json:"cycle_time_days"`
	Lognormal         Lognormal  `json:"lognormal"`
	WipCount          int        `json:"wip_count"`
	ThroughputPerWeek float64    `json:"throughput_per_week"`
}

type CycleTimes struct {
	P50  float64 `json:"p50"`
	P85  float64 `json:"p85"`
	Mean float64 `json:"mean"`
}

type Lognormal struct {
	Mu    float64 `json:"mu"`
	Sigma float64 `json:"sigma"`
}

// Edge mirrors one entry of edges.json: a directed cross-pod dependency.
type Edge struct {
	From  string `json:"from"`
	To    string `json:"to"`
	Count int    `json:"count"`
}

// podAliases maps Jira pod-field spellings to the org directory's pod names.
var podAliases = map[string]string{"Moose Factory": "MooseFactory"}

func podOf(raw string) string {
	if a, ok := podAliases[raw]; ok {
		return a
	}
	return raw
}

const (
	maxCycleDays = 180.0
	excludeEpic  = "Epic"
	excludeEpic2 = "Parent Epic"
	weeks180     = 26.0 // ~180 days, matching aggregate.py's throughput window
)

// Aggregate turns issues into per-pod stats and cross-pod edges, reproducing
// scripts/aggregate.py: epics and >180d backlog artifacts are dropped from
// cycle time, the rest winsorized at the pod's p95; edges are deduped
// (blocker,blocked) pairs across different pods. wipMode picks how epics vs.
// their children are counted toward wip (see WipModeLeaf / WipModeEpicOrParentless).
func Aggregate(issues []Issue, wipMode string) (map[string]PodStat, []Edge) {
	podByKey := map[string]string{}
	for _, it := range issues {
		if p := podOf(it.Pod); p != "" {
			podByKey[it.Key] = p
		}
	}

	// dedup (blocker, blocked) pairs -> cross-pod edge counts
	pairs := map[[2]string]bool{}
	for _, it := range issues {
		for _, k := range it.Blocks {
			pairs[[2]string{it.Key, k}] = true
		}
		for _, k := range it.BlockedBy {
			pairs[[2]string{k, it.Key}] = true
		}
	}
	edgeCount := map[[2]string]int{}
	for pair := range pairs {
		pa, pb := podByKey[pair[0]], podByKey[pair[1]]
		if pa == "" || pb == "" || pa == pb {
			continue
		}
		edgeCount[[2]string{pa, pb}]++
	}
	edges := make([]Edge, 0, len(edgeCount))
	for k, c := range edgeCount {
		edges = append(edges, Edge{From: k[0], To: k[1], Count: c})
	}
	// deterministic: count desc, then names
	sort.Slice(edges, func(i, j int) bool {
		if edges[i].Count != edges[j].Count {
			return edges[i].Count > edges[j].Count
		}
		if edges[i].From != edges[j].From {
			return edges[i].From < edges[j].From
		}
		return edges[i].To < edges[j].To
	})

	// cycle-time samples per pod
	byPod := map[string][]float64{}
	for _, it := range issues {
		p := podOf(it.Pod)
		if p == "" || it.Resolved == nil || it.Created.IsZero() {
			continue
		}
		if it.IssueType == excludeEpic || it.IssueType == excludeEpic2 {
			continue
		}
		days := it.Resolved.Sub(it.Created).Hours() / 24
		if days < 0.25 {
			days = 0.25
		}
		if days > maxCycleDays {
			continue
		}
		byPod[p] = append(byPod[p], days)
	}
	// winsorize each pod at its p95 (index int(0.95*(n-1)), matching aggregate.py)
	for p, xs := range byPod {
		sort.Float64s(xs)
		cap := xs[int(0.95*float64(len(xs)-1))]
		for i := range xs {
			if xs[i] > cap {
				xs[i] = cap
			}
		}
		byPod[p] = xs
	}

	// wip = actively in progress (status category "indeterminate"), matching
	// the drill-down's definition (server/db/snapshotquery.go WipPage/WipSummary)
	// — an unresolved backlog issue is not wip just because it's not done yet.
	wipByPod := map[string]int{}
	for _, it := range issues {
		if it.Resolved != nil || it.StatusCat != "indeterminate" {
			continue
		}
		isEpic := it.IssueType == excludeEpic || it.IssueType == excludeEpic2
		counts := !isEpic // WipModeLeaf: every non-epic issue
		if wipMode == WipModeEpicOrParentless {
			counts = isEpic || it.ParentKey == ""
		}
		if !counts {
			continue
		}
		if p := podOf(it.Pod); p != "" {
			wipByPod[p]++
		}
	}

	stats := map[string]PodStat{}
	for p, xs := range byPod {
		logs := make([]float64, len(xs))
		var sum, lsum float64
		for i, x := range xs {
			logs[i] = math.Log(x)
			lsum += logs[i]
			sum += x
		}
		mu := lsum / float64(len(logs))
		var varSum float64
		for _, l := range logs {
			varSum += (l - mu) * (l - mu)
		}
		sigma := math.Sqrt(varSum / float64(len(logs)))
		if sigma == 0 {
			sigma = 0.1
		}
		stats[p] = PodStat{
			ResolvedCount180d: len(xs),
			CycleTimeDays:     CycleTimes{P50: round2(pct(xs, 50)), P85: round2(pct(xs, 85)), Mean: round2(sum / float64(len(xs)))},
			Lognormal:         Lognormal{Mu: round4(mu), Sigma: round4(sigma)},
			WipCount:          wipByPod[p],
			ThroughputPerWeek: round2(float64(len(xs)) / weeks180),
		}
	}
	// pods with WIP but nothing resolved still deserve an entry
	for p, w := range wipByPod {
		if _, ok := stats[p]; ok {
			continue
		}
		stats[p] = PodStat{
			ResolvedCount180d: 0,
			CycleTimeDays:     CycleTimes{P50: 7.0, P85: 17.0, Mean: 10.0},
			Lognormal:         Lognormal{Mu: math.Log(7), Sigma: 0.9},
			WipCount:          w,
			ThroughputPerWeek: 0.0,
		}
	}
	return stats, edges
}

// pct returns the linearly-interpolated q-th percentile of xs (xs is sorted).
func pct(xs []float64, q float64) float64 {
	if len(xs) == 0 {
		return 0
	}
	idx := (q / 100) * float64(len(xs)-1)
	lo, hi := math.Floor(idx), math.Ceil(idx)
	return xs[int(lo)] + (xs[int(hi)]-xs[int(lo)])*(idx-lo)
}

func round2(x float64) float64 { return math.Round(x*100) / 100 }
func round4(x float64) float64 { return math.Round(x*10000) / 10000 }
